dfu/boot: rework using newly introduced MCUBOOT_BOOTUTIL library

nvlsianpu opened the pull request:
depends on https://github.com/mcu-tools/mcuboot/pull/892

As some code are defined here and there it becomes problem when module is compiled within MCUboot.
It is possible to mitigate problem using #fdefry - but this is temporary hack.

aim to supersede #30060

fixes #29963

- [x] takes needs of https://github.com/zephyrproject-rtos/zephyr/pull/29669 into account

**step needed for merge this PR**

get the Zephyr's release-team member who can assist with merges sequence smoothly

1. [x]  1. PR approved, 
2. [x] 2.  merge https://github.com/mcu-tools/mcuboot/pull/892 (is approved, ready to be merged)
3. [x] 3. create upmereg PR to https://github.com/zephyrproject-rtos/mcuboot
4. [x] 4. merge the upmerge PR
5. [x] 5. update mcuboot SHA and merge this PR.

nvlsianpu committed:
modules: introduce MCUBOOT_BOOTUTIL module

Some parts of subsys/dfu/boot code are re-implementations of
what is implemented in the MCUBoot repository.

Mcuboot's repository already provide implementation of function
required for application for interact with the MCUboot.

This patch introduces new MCUBOOT_BOOTUTIL module which covers
common code which is used in the bootloader and the chainnloaded
application.

dfu/boot: use MCUBoot's source code

Module was reworked so it start using MCUBoot's
bootutil_public API instead of copied code.

Reworked boot_is_img_confirmed() used MCUBoot's API
for determine image_ok flag.

mcuboot_shell switchd to use MCUboot's boot_read_swap_state_by_id()
This is MCUBoot function, use it for avoid linking conflict.

test/subsys/mcuboot: fix `test_write_confirm`

dfu/boot library was reworked so it uses MCUboot's bootutil_public
library whenever it can.
The library required that image was marked as copy-done before it
can be pending.
This patch adds such mark which fixes the test.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented on modules/Kconfig.mcuboot_bootutil:
```diff
@@ -0,0 +1,17 @@
+	  requret by the chainnloaded application and the MCUBoot.
+
+if !MCUBOOT && MCUBOOT_BOOTUTIL_LIB
+module = MCUBOOT_BOOTUTIL
```
@tejlmand Is there possible to move this declaration to mcuboot repo. I have been trying to do so but couldn't menage to move it - even i was able to include a new Kconfig, it was not visible in kconfig processing results.

nvlsianpu replied:
I found another way, I made this able to be disabled by MCUBoot.

tejlmand replied:
I see that :+1: 
Sorry for the delay for getting to this.

I have placed a comment in the Kconfig for you to consider, though it's not blocking.

de-nordic commented on modules/Kconfig.mcuboot_bootutil:
```diff
@@ -0,0 +1,17 @@
+	bool "mcuboot utility library"
+	help
+	  Enable mcuboot utility library which implements functions
+	  requret by the chainnloaded application and the MCUBoot.
```
`requert`->`required`, I think.

utzig commented on include/dfu/mcuboot.h:
```diff
@@ -17,26 +17,57 @@
+#endif
 
+#ifdef BOOT_SWAP_TYPE_REVERT
+#if BOOT_SWAP_TYPE_REVERT != 4  /*ensure the same definition in MCUBoot */
```
Is this really supposed to check for `BOOT_SWAP_TYPE_REVERT` and not `BOOT_SWAP_TYPE_FAIL`?

utzig commented on modules/Kconfig.mcuboot_bootutil:
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+
```
I believe you could remove the extra empty line here.

utzig commented on modules/Kconfig.mcuboot_bootutil:
```diff
@@ -0,0 +1,24 @@
+
+
+menuconfig MCUBOOT_BOOTUTIL_LIB
+	bool "mcuboot utility library"
```
`MCUboot`

utzig commented on include/dfu/mcuboot.h:
```diff
@@ -17,26 +17,57 @@
 #endif
 
+#ifdef BOOT_SWAP_TYPE_NONE
+#if BOOT_SWAP_TYPE_NONE != 1 /*ensure the same definition in MCUBoot */
```
The proper case would be `MCUboot` :-)

nvlsianpu replied:
Do you mean BOOT_SWAP_TYPE_NONE -> MCUBOOT_SWAP_TYPE_NONE.
This will impose on users (mcumgr, updatehub...), that why I keep naming.

utzig replied:
Hehe, no I mean in the comment you should change "MCUBoot" to "MCUboot"

utzig commented on tests/subsys/dfu/mcuboot/src/main.c:
```diff
@@ -117,7 +118,12 @@ void test_write_confirm(void)
 
-	zassert(boot_write_img_confirmed() == 0, "pass", "fail");
+	/* set copy-done flag */
+	ret = flash_area_write(fa, fa->fa_size - 32, &flag, sizeof(flag));
```
This is a bit confusing to me, you're writing to the correct location but you're writing 4 bytes (`uint32_t`) instead of the usual 8 bytes that MCUboot uses for a flag, and even if it works (with the 0x01 at the right place) it will only work for one specific endianness.

nvlsianpu replied:
I believe this test-sute entirely written like that (assues 32 bits write access and lite-endianes).
I can rework it to be more widely, but see this as another PR.
Is that Ok for you?

utzig replied:
It is acceptable, but I don't see why not just fixing it properly since it's two lines of change.

nvlsianpu replied:
ok. done

utzig commented on include/dfu/mcuboot.h:
```diff
@@ -181,6 +220,17 @@ int boot_request_upgrade(int permanent);
+ * @return 0 on success, negative errno code on fail.
+ */
+int mcuboot_read_swap_state_by_id(int flash_area_id,
+				  struct mcuboot_swap_state *state);
```
Is there a rationale for adding this function and `struct mcuboot_swap_state` instead of just using `boot_read_swap_state_by_id` and `struct boot_swap_state`?

nvlsianpu replied:
boot_read_swap_state_by_id() can't be made zephyr API. It can only be used by module which made a module shim (as it links with MCUBOOT_BOOTUTIL) (by convention). For the MCUBoot shell  i can link it with the library as well, and left behind this API.
Do you think that this API will be rather not useful for other?

nvlsianpu replied:
@utzig ^^?

mniestroj replied:
Adding `image_num` would be a short term fix. We should not blindly cast one structure to the other, because that is error prone. Somebody will change order of fields or add new information to `struct boot_swap_state` and we will be screwed once again.

So we should either:
- add helper `struct boot_swap_state` variable and copy each structure field separately,
- link `MCUBOOT_BOOTUTIL` to mcuboot shell or other modules that use mcuboot public API.

nvlsianpu replied:
> - link MCUBOOT_BOOTUTIL to mcuboot shell or other modules that use mcuboot public API.

Ok, I will go with that

nvlsianpu replied:
done

nvlsianpu commented on include/dfu/mcuboot.h:
```diff
@@ -17,26 +17,57 @@
+#endif
 
+#ifdef BOOT_SWAP_TYPE_REVERT
+#if BOOT_SWAP_TYPE_REVERT != 4  /*ensure the same definition in MCUBoot */
```
From running application context I see expected are: NONE and TEST. Should I keep only these?

nvlsianpu replied:
@utzig  ^^?

utzig replied:
Let me copy/paste here then:

```
#ifdef BOOT_SWAP_TYPE_REVERT
#if BOOT_SWAP_TYPE_REVERT != 4  /*ensure the same definition in MCUboot */
#error "definition incompatible"
#endif
/** Swap failed because image to be run is not valid */
#define BOOT_SWAP_TYPE_FAIL     5
#endif
```

I don't understand why you need to check for `BOOT_SWAP_TYPE_REVERT` while defining `BOOT_SWAP_TYPE_FAIL`, is there a rationale?

nvlsianpu replied:
missed that, was a typo

tejlmand commented on modules/Kconfig.mcuboot_bootutil:
```diff
@@ -0,0 +1,23 @@
+
+#hiden option for disabling module-own log configuration
+#while building MCUBoot bootloader
+config DISABLE_MCUBOOT_BOOTUTIL_LIB_OWN_LOG
+	bool
```
Nothing wrong, but generally I think a `MCUBOOT_BOOTUTIL_LIB_LOG=y` is nicer and easier to read.
```suggestion
config MCUBOOT_BOOTUTIL_LIB_LOG
	bool
	default y
```
and then in mcuboot do:
```
config MCUBOOT_BOOTUTIL_LIB_LOG
	bool
	default n
```

this is one of the few places where a `default n` is actually meaningful, and will work when placed in https://github.com/zephyrproject-rtos/mcuboot/blob/master/boot/zephyr/Kconfig before the `Kconfig.zephyr` is sourced:
https://github.com/zephyrproject-rtos/mcuboot/blob/6e3825f1309b34d3f5e5e1be8dd522cfe92e044b/boot/zephyr/Kconfig#L574

Ref: https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#redundant-defaults
> The one case where `default n/default ""` is not redundant is when defining a symbol in multiple locations and wanting to override e.g. a default y on a later definition.

nvlsianpu replied:
reworked

nvlsianpu commented:
@utzig can you revisit?

mniestroj commented:
Currently this PR breaks git bisectability: all commits (except the last one that touches tests) are needed in order to build successfully.

mniestroj commented:
@tejlmand What is the convention about choosing cmake library name with `zephyr.*library_named()`? I see some places use lowercase names, others uppercase. Shouldn't we decide on one or the other?

nvlsianpu commented:
> Currently this PR breaks git bisectability: all commits (except the last one that touches tests) are needed in order to build successfully.

I will squash patches once PR changes will be accepted.

nvlsianpu committed:
tests/subsys/dfu/mcuboot: support mcuboot max alignment

MCUboot support flash write-bock-size up to BOOT_MAX_ALIGN.
This patch takes this into account.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

de-nordic reviewed (approved):
Looks OK.

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -469,221 +149,58 @@ int boot_read_bank_header(uint8_t area_id,
-bool boot_is_img_confirmed(void)
-{
-	return boot_image_ok_read(FLASH_AREA_IMAGE_PRIMARY) == BOOT_FLAG_SET;
+	return flag_val == BOOT_FLAG_SET;
```
Reworking this would be an API change - so, yes not in this PR.

nvlsianpu commented:
> Currently this PR breaks git bisectability: all commits (except the last one that touches tests) are needed in order to build successfully.

@mniestroj Squashed

mniestroj commented on include/dfu/mcuboot.h:
```diff
@@ -17,26 +17,57 @@
 #define BOOT_MAX_ALIGN		8
+#ifndef BOOT_MAGIC_SZ
 #define BOOT_MAGIC_SZ		16
+#endif
```
`BOOT_MAGIC_SZ` is not needed anymore.

`BOOT_MAX_ALIGN` as well. The only place where you need it is unit test that you have updated. I think MCUBOOT_BOOTUTIL could be linked to unit test to provide that symbol.

nvlsianpu replied:
both definition are needed in mcumgr and dfu/img_utill/flash_img.c under BOOT_TRAILER_IMG_STATUS_OFFS macro.

As you mention these definition can be removed once that code will be linked with MCUBOOT_BOOTUTIL.

mniestroj commented on include/dfu/mcuboot.h:
```diff
@@ -17,26 +17,57 @@
 #endif
 
+#ifdef BOOT_SWAP_TYPE_NONE
+#if BOOT_SWAP_TYPE_NONE != 1 /*ensure the same definition in MCUboot */
```
The only place where `BOOT_SWAP_TYPE_` are needed (but are not provided by MCUBOOT_BOOTUTIL) is mcumgr port. We don't need to do it in this PR, but IMO mcumgr should link to MCUBOOT_BOOTUTIL cmake target to get those macros. After doing so, BOOT_SWAP_TYPE_ macros could be thrown away from this header. This is the last duplicated code from mcuboot I guess.

nvlsianpu replied:
I prefer to do that in another PR.

nvlsianpu commented:
@mniestroj Regard next step for eliminate copied definitions (so linking test, mcumgr, img_utill to MCUBOOT_BOOTUTIL library)- I will do that in the next PR which will open immediately after this will be merge. I think this PR is heavy enough - I prefer smaller chunks of patches.

nvlsianpu commented:
mcuboot counterpart PR https://github.com/zephyrproject-rtos/mcuboot/pull/42
